002-tidyr-stringr.Rmd

```diff
@@ -0,0 +1,228 @@
+    There's something nice about an s verb for tidyr since it feels "string-y".
+
+    For the purpose of completeness, other unused split/separate synonyms: fracture, cleave, snap, splinter, divide, sever, rend, dissect.
+
+### Eliminate tidyr functions
+
+Before we continue on to other design considerations, it's worth double checking that we can't eliminate the tidyr function in favour of using dplyr and stringr directly.
+This seemed like the "obvious" solution to me at least twice, so I wanted to make sure why it doesn't work.
+
+To make the discussion concrete I'll start with quick and dirty implementations of `str_separate_by_wider()` and `str_separate_by_longer()`:
+
+```{r}
+str_separate_by_wider <- function(x, pattern, into) {
+  mat <- str_split_fixed(x, pattern, length(into))
+  colnames(mat) <- into
+  tibble::as_tibble(mat)
+}
+str_separate_by_longer <- function(x, pattern) {
+  str_split(x, pattern)
+}
+
+str_separate_by_wider(c("a-b", "c-d"), "-", c("first", "second"))
+str_separate_by_longer(c("a-b", "c-d-e"), "-")
+```
+
+We can use `str_separate_by_wider()` directly with `mutate()` to achieve the same affect as `separate()`:
+
+```{r}
+library(dplyr, warn.conflicts = FALSE)
+
+df <- tibble(x = c("a-b", "c-d"))
+df %>% separate(x, c("first", "second"), "-")
+
+df %>% 
+  mutate(
+    str_separate_by_wider(x, "-", c("first", "second")), 
+    .keep = "unused"
+  )
+```
+
+This works because `mutate()` will create multiple columns if an unnamed calculation returns a data frame.
+
+However, there's no straightforward way to use `str_separate_by_longer()` to create multiple columns.
+The simplest pure dplyr approach is probably to use `rowwise():`
+
+```{r}
+df %>%
+  rowwise() %>% 
+  summarise(str_separate_by_longer(x, "-")[[1]])
+```
+
+But it's quite hard to explain why this works, and the use of `summarise()` means that you need to explicitly list any variables that you want to leave as is.
+This possibly speaks to a missing verb in dplyr (e.g. <https://github.com/tidyverse/dplyr/issues/5874>) but it's also possible that creating rows is fundamentally a tidyr operation.
+Either way, there's no easy fix in sight, and it seems unappealing to force folks to learn sophisticated dplyr idioms in order to tackle a routine data tidying task.
+
+### Arguments
+
+-   `str_separate_*_*()`: first argument is character vector.
+
+-   `str_*_*_wider()` needs a `col_names` argument and arguments that controls what happens if there are too many/too few pieces.
+    Probably needs `names_sep` for consistency with `unnest_wider()`.
+
+```{=html}
+<!-- -->
+```
+-   `str_separate_group_wider()` has a few options for specifying names:
+
+    -   Named groups `"(?<type>.*)|(?<fruit>orange|apple)`
+
+    -   Named vector: `c(type = ".*", "|", fruit = "orange|apple"))`
+
+-   Should `convert` continue to exist?
+    Should it use `type.convert()` or `readr::type_convert()`?
+    How to override imputed types?
```

Comment:
I much prefer the finer control offered by `tidyr::pivot_longer(values_transform = )`. Could be `into_transform =`.

Reply:
Or just `transform` to match `unnest_wider()`?